include/alpaka/mem/buf/SetKernel.hpp

```diff
@@ -0,0 +1,80 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/mem/buf/Traits.hpp>
+#include <alpaka/idx/Traits.hpp>
+#include <alpaka/idx/Accessors.hpp>
+#include <alpaka/idx/MapIdx.hpp>
+
+namespace alpaka
+{
+    namespace mem
+    {
+        namespace view
+        {
+            //#############################################################################
+            //! any device ND memory set kernel.
+            class MemSetKernel
+            {
+            public:
+                //-----------------------------------------------------------------------------
+                //! The kernel entry point.
+                //!
+                //! All but the last element of threadElemExtent must be one.
+                //!
+                //! \tparam TAcc The accelerator environment to be executed on.
+                //! \tparam TElem element type.
+                //! \tparam TExtent extent type.
+                //! \param acc The accelerator to be executed on.
+                //! \param val value to set.
+                //! \param dst target mem ptr.
+                //! \param extent area to set.
+                ALPAKA_NO_HOST_ACC_WARNING
+                template<
+                    typename TAcc,
+                    typename TElem,
+                    typename TExtent,
+                    typename TPitch>
+                ALPAKA_FN_ACC auto operator()(
+                    TAcc const & acc,
+                    TElem const val,
+                    TElem * dst,
+                    TExtent extent,
+                    TPitch pitch) const
+                -> void
+                {
+                    using Idx = typename idx::traits::IdxType<TExtent>::type;
+                    auto const gridThreadIdx(alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc));
+                    auto const threadElemExtent(alpaka::workdiv::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc));
+                    auto const idxThreadFirstElem = idx::getIdxThreadFirstElem(acc, gridThreadIdx, threadElemExtent);
+                    auto idx = idx::mapIdxPitch<1u, dim::Dim<TAcc>::value>(idxThreadFirstElem, pitch)[0];
+                    constexpr auto lastDim = dim::Dim<TAcc>::value - 1;
+                    const auto lastIdx = idx +
+                        std::min(threadElemExtent[lastDim], static_cast<Idx>(extent[lastDim]-idxThreadFirstElem[lastDim]));
+
+                    if ([&idxThreadFirstElem, &extent](){
+                            for(auto i = 0u; i < dim::Dim<TAcc>::value; ++i)
+                                if(idxThreadFirstElem[i] >= extent[i])
+                                    return false;
+                            return true;
+                        }())
+                    {
+                        // assuming elements = {1,1,...,1,n}
+                        for(; idx<lastIdx; ++idx)
+                        {
+                            dst[idx] = val;
```

Comment:
Given my example where the element size is not a multiple of the pitch, it is invalid to index into a multidimensional buffer outside of the first row.
We can not use `dst` from the outside here directly. We have to calculate the base pointer for each thread manually.

Reply:
You can find the corresponding code here: https://github.com/alpaka-group/alpaka/blob/develop/test/common/include/alpaka/test/mem/view/Iterator.hpp#L141-L153
`m_pitchBytes` -> `pitch`
`currentIdxDimx` -> `idxThreadFirstElem`

Reply:
Why is this iterator part of the test tree and not of alpaka? I could implement `SetKernel` using this iterator instead of MapIdxPitch. Could we move this intro the alpaka?

Reply:
Specifically in the case of `SetKernel` the current implementation does work because memset always sets memory byte-wise. The limit checking the row-extent in line 60,61 ensures that it does not write into the padding space.

Reply:
The Iterator is not part of alpaka because nobody moved it there and it was only used for testing yet.
You can move it to `alpaka/mem/view` and add a short test. This would allow to remove `mapIdxPitchBytes`.